internal/pkg/slicer/rowsreader/reader.go

```diff
@@ -216,10 +216,9 @@ func (r *Reader) openSlice(path string) (*sliceReadCloser, error) {
 		if gzipReader, err := r.gzipReaders.ReaderFrom(out.Reader); err == nil {
 			out.Reader = gzipReader
 			out.Closers.
-				Append(gzipReader.Close).
 				Append(func() error {
-					r.gzipReaders.Put(gzipReader)
-					return nil
+					defer r.gzipReaders.Put(gzipReader)
+					return gzipReader.Close()
```

Comment:
I didn't realize that callbacks are executed from below, in LIFO order.

So the reader was **first returned to the pool and then closed.** Normally it would throw a `data race` error in tests. But as I wrote in the PRintroduction, another mechanism in the code ensures that the problem does not actually occur.

---------

I used the `defer' statement (similar to the PHP `finally`), because it is executed even in case of panic. Even if we don't currently catch panic anywhere, we never ignore it and program is always stopped (it is fatal error).